### DISCUSSION TITLE
Room for improving installation speed?

### DISCUSSION THREAD

**Discussion by privatenumber** [semantic-release/semantic-release#2885] [Ideas]
### Current behavior

I've been an avid user of semantic-release for a few years now and I really appreciate the amazing work done by the maintainers and contributors to make this tool so valuable for the community.

I noticed that the [installation size for v21.0.7 amounts to about 32MB](https://pkg-size.dev/semantic-release). I don't think plugins are included in this figure but I was surprised to learn that the bundled `semantic-release` itself is only 750KB minified. This suggests that there might be a considerable amount of unused code being installed.

Since semantic-release is primarily used on CI, I think minimizing installation time could save costs for many users. Additionally, it could lead to a snappier experience when used with `npx`.

One point of improvement that stands out in the installation size is `@octokit/openapi-types`, accounting for 13% (4.2MB). As this package contains only types and isn't utilized at runtime, it may not need to be installed.

Additionally, I've noticed a few other packages like `npm`, `uglify-js`, and `handlebars`, which makes me wonder why and if they're necessary.

### Expected behavior

Only for code that is used to be installed.

### `semantic-release` version

21.0.7

### CI environment

WebContainers

### `semantic-release` configuration

N/A

### CI logs

N/A

**Comment by travi** [comment]
appreciate the analysis, but i'm unsure how some of the details that you are highlighting apply to the project. we certainly want to avoid including unnecessary dependencies that our consumers need to download without value, but all of the packages that you mentioned are either needed by our core or plugins directly, or are dependencies of our necessary dependencies. 

> the bundled `semantic-release` itself is only 750KB minified

additionally, i'm confused by this figure since we do not bundle or minify our package. semantic-release does not run in a browser context and wouldnt be a supported use case, so minification doesnt really apply. 

> One point of improvement that stands out in the installation size is `@octokit/openapi-types`, accounting for 13% (4.2MB). As this package contains only types and isn't utilized at runtime, it may not need to be installed.

while this may be a valid point, this is out of our control directly. as you can see, it is a transitive dependency of `@octokit/core`. it would be better to work through that concern with the octokit team directly rather than through us.

```
$ npm ls @octokit/openapi-types
semantic-release@0.0.0-development 
└─┬ @semantic-release/github@9.0.4
  └─┬ @octokit/core@5.0.0
    └─┬ @octokit/types@11.1.0
      └── @octokit/openapi-types@18.0.0
```

**Reply by privatenumber** [reply]
> additionally, i'm confused by this figure since we do not bundle or minify our package. semantic-release does not run in a browser context and wouldnt be a supported use case, so minification doesnt really apply.

Bundling isn't an optimization limited to browsers (e.g. every major bundler has a Node.js target). It can be leveraged here to reduce the install size dramatically.

> while this may be a valid point, this is out of our control directly. as you can see, it is a transitive dependency of @octokit/core. it would be better to work through that concern with the octokit team directly rather than through us.

`@octokit/core` likely offers types for its own usage. But if it's being passed onto your users, you can remove this by bundling.
